### DISCUSSION TITLE
FocusManager Error

### DISCUSSION THREAD

**Discussion by xiaoliangde** [AvaloniaUI/Avalonia#17681] [Q&A]
### Describe the bug

`public TreeItemIconTextBase<TNode> WithEditText(bool splitExtension)
{
    if (_hadInitEditMode) return this;
    _hadInitEditMode = true;

    TextShowControl.PointerPressed += (sender, args) => _pressedTime = DateTime.Now.ToTimestamp();
    TextShowControl.PointerReleased += (sender, args) => { if (DateTime.Now.ToTimestamp() - _pressedTime > TopLevel.GetTopLevel(this).PlatformSettings.GetDoubleTapTime(PointerType.Mouse).TotalMilliseconds) TextShowControl.IsVisible = !TextShowControl.IsVisible; };
    _editValue = SplitFileName(TextShowControl.Text,splitExtension);

    var cTbEdit = new TextBoxExFocus()
    {
        Text = _editValue.NameWithoutExtension,
        Margin = new Thickness(5, 0, 0, 0),
        FontSize = 14,
        VerticalAlignment = VerticalAlignment.Center,
        VerticalContentAlignment = VerticalAlignment.Center,
        MaxWidth = 220,
        MinWidth = 60,
        MaxLength = 48,
        Padding = new Thickness(0),
        Background = Brushes.Transparent,
    };
    // using var once = TextControl.Notify(nameof(TextControl.Text), cTbEdit, nameof(cTbEdit.Text), o => (o is string val) ? SplitFileName(val, splitExtension).NameWithoutExtension : string.Empty, o => $"{o}{_editValue.Extension}");
    TextShowControl.Notify(nameof(TextShowControl.IsVisible), cTbEdit, nameof(cTbEdit.IsVisible), 
        s=>
        {
            cTbEdit.IsVisible = !TextShowControl.IsVisible;
            if (!cTbEdit.IsVisible) return;
            cTbEdit.FocusAndSelect();//If Use Focus() will be incorrect.
        },
        t=>TextShowControl.IsVisible = !cTbEdit.IsVisible);
    cTbEdit.Text = _editValue.NameWithoutExtension;
    cTbEdit.PointerReleased += (o, args) => args.Handled = true;
    cTbEdit.PointerPressed += (o, args) => args.Handled = true;
    cTbEdit.Tapped += (o, args) => args.Handled = true;
    cTbEdit.DoubleTapped += (o, args) => args.Handled = true;
    cTbEdit.LostFocus += (s, args) => TryFinishEdit(cTbEdit);
    cTbEdit.KeyDown += (s, args) =>
    {
        if (args.Key != Key.Enter) return;
        args.Handled = true;
        TryFinishEdit(cTbEdit);
    };

    RowPanel.Children.Add(cTbEdit);
    DockPanel.SetDock(cTbEdit, Dock.Left);
    return this;
}`

the Control.Focus() can't be correct result（here "cTbEdit.FocusAndSelect()" Impl by FocuseManage Internal Method and it can be correct）.
It's incorrect usually come by FlyoutMenu("Rename" Item).
![image](https://github.com/user-attachments/assets/53481816-d487-420e-ab60-6108083b9b32)
![image](https://github.com/user-attachments/assets/0cf8d328-3402-4065-9064-57d1e6344b9c)
...But But FocuseManage Internal Method or Directly Call Control.Focus() Not work well by "OnLoad" or "OnAttachXXTree" 
——
![image](https://github.com/user-attachments/assets/91ab41b4-4ede-441f-9a9d-e33e6cddfe92)
Except add the timer to delay for test，I don't know how to Focus after It be Created。Now it had be click once to focus.
—— after manual click
![image](https://github.com/user-attachments/assets/353ef9bc-289f-4b79-a5c2-1d18155495fe)

### To Reproduce

"Focus()"（even User Internal FocusManager Method） can't be work onLoad or OnAttachXXTree

### Avalonia version

11.0.6

**Comment by xiaoliangde** [comment]
public MainViewEditor()
        {
            EnvSetting.Current.IsEditorMode = true;
            Focusable = true;
            IEditor.Current = this;
            InitializeComponent();
            _cLeftContent.Content = LeftPanel = new LeftPanel();

          

            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.NextTabItemCanvas));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.LastTabItemCanvas));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.CloseCurrentTabItemCanvas));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.CloseAllTabItemCanvas));

            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.ToggleLeftPanel));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.ToggleRightPanel));

            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.NewProject));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.OpenProject));
            KeyBindings.Add(new KeyBinding().SetCmdAndKey(CommandsEditorGlobal.Instance.SaveAllPages));
        }

        protected override void OnAttachedToLogicalTree(LogicalTreeAttachmentEventArgs e)
        {
            base.OnAttachedToLogicalTree(e);
            this.Focus();//Not Correct
            this.SetFocusedElement(NavigationMethod.Directional, KeyModifiers.None);//Not Correct Too
        }
        
        —— The MainWindow.Content UI Code. 
        UI Not correct and KeyBinding also

**Comment by timunie** [comment]
You may want to use Dispatcher.UiThread.Post to move focus. Priority Loaded.

**Comment by xiaoliangde** [comment]
> You may want to use Dispatcher.UiThread.Post to move focus. Priority Loaded.

Good idea for using =>Dispatcher.UiThread.Post. It looks like correct.the shortkey can be used.
thanks.
